aws/s3.go

```diff
@@ -253,10 +255,54 @@ func getBucketInfo(svc *s3.S3, bucket *s3.Bucket, excludeAfter time.Time, region
 		return
 	}
 
+	// Check if the bucket matches config file rules
+	if !shouldIncludeBucket(bucketData.Name, configObj.S3.IncludeRule.NamesRE, configObj.S3.ExcludeRule.NamesRE) {
+		bucketData.InvalidReason = "Filtered by config file rules"
+		bucketCh <- &bucketData
+		return
+	}
+
 	bucketData.IsValid = true
 	bucketCh <- &bucketData
 }
 
+func shouldIncludeBucket(bucketName string, includeNamesREList []*regexp.Regexp, excludeNamesREList []*regexp.Regexp) bool {
+	shouldInclude := false
+
+	if len(includeNamesREList) > 0 {
+		if includeBucketByREList(bucketName, includeNamesREList) {
+			shouldInclude = excludeBucketByREList(bucketName, excludeNamesREList)
+		}
+	} else if len(excludeNamesREList) > 0 {
+		shouldInclude = excludeBucketByREList(bucketName, excludeNamesREList)
+	} else {
+		shouldInclude = true
+	}
+
+	return shouldInclude
```

Comment:
NIT: I think this logic can be simplified down to:

```go
if len(includeNamesREList) > 0 && includeBucketByREList(bucketName, includeNamesREList) {
  return excludeBucketByREList(bucketName, excludeNamesREList)
} else if len(excludeNamesREList) > 0 {
  return excludeBucketByREList(bucketName, excludeNamesREList)
} else {
  return true
}

Reply:
This introduces a bug, actually, so not implementing the change. I will explain with data why this doesn't work.

Reply:
So the reason the simplification is inaccurate is that, when `includeBucketByREList` returns false, we shouldn't ever check that same bucket with `excludeBucketByREList`. (I need to rename these methods, both of them are including buckets, but by either include regex lists or exclude regex lists.)

If the exclude rule ends up including a bucket that the include rule already excluded, that's a bug. So this expanded logic is for that purpose.